core/engine.go

```diff
@@ -106,128 +106,182 @@ func NewEngine(ex lib.ExecutionScheduler, o lib.Options, logger *logrus.Logger)
 	return e, nil
 }
 
-// Init is used to initialize the execution scheduler. That's a costly operation, since it
-// initializes all of the planned VUs and could potentially take a long time.
-func (e *Engine) Init(ctx context.Context) error {
-	return e.ExecutionScheduler.Init(ctx, e.Samples)
-}
-
-func (e *Engine) setRunStatus(status lib.RunStatus) {
-	for _, c := range e.Collectors {
-		c.SetRunStatus(status)
+// Init is used to initialize the execution scheduler and all metrics processing
+// in the engine. The first is a costly operation, since it initializes all of
+// the planned VUs and could potentially take a long time. It either returns an
+// error immediately, or it returns test run() and wait() functions.
+//
+// Things to note:
+//  - The first lambda, Run(), synchronously executes the actual load test.
+//  - It can be prematurely aborted by cancelling the runCtx - this won't stop
+//    the metrics collection by the Engine.
+//  - Stopping the metrics collection can be done at any time after Run() has
+//    returned by cancelling the globalCtx
+//  - The second returned lambda can be used to wait for that process to finish.
+func (e *Engine) Init(globalCtx, runCtx context.Context) (run func() error, wait func(), err error) {
```

Comment:
There is a `TODO` in the `Engine` about splitting the lambdas into their own sub-object, an `EngineSession` of sorts. But I left that for a future refactoring, since this took too long... :disappointed: